### PR TITLE
Extract Network and Interface from utils

### PR DESCRIPTION
**Pull request by Hoverbear**
We would like to be able to use this code in benchmarks (#109) and other tests but currently there is no way to use `Network` or `Interface` from non-integration test code.

This introduces a `harness` feature which exposes `harness::{Interface, Network}`.

There is some rationale to how this was done:

* We cannot have a module in `tests/` that is built for benchmarking, so they were moved into a test harness module.
* We cannot use `#[cfg(test)]` or `#[cfg(debug_assertions)]` to have this only build in test or development, since bench marking builds in release. So we have to make a feature flag.
* We cannot control which features are used during `cargo test` or `cargo bench` by default without altering all defaults (such as what users get when they import Raft). So we must make this a default feature.

Benefits:
* Opens the harness to users in their application code for exploring or testing.
* Opens the harness to benchmarking and unit tests.
* Allows `Network` and `Interface` to have their own modules as per #125 .

Detriments:
* Users may get confused, but the documentation is quite explicit that it is simulator code.
* It will slightly increase non `cargo test` compile time. The impact should be minimal.
* Users may wish to disable the harness, which would require manual work on their end.

### CODE REVIEW

**Comment by hicqu** [comment]
Make `harness` as a default feature is not very friendly for user. For example, TiKV needs to disable the feature when declar the raft dependency. So how about make `harness` as a sub crate?

**Comment by Hoverbear** [comment]
@hicqu I think "need" is a very strong word, the only reason for TiKV to disable the feature is if we were worried a few milliseconds of compile time.

I think the idea of using a component is very interesting. I will explore this. :)

**Comment by Hoverbear** [comment]
@hicqu Your idea was very good I think! PTAL again?

**Comment by hicqu** [comment]
LGTM. Please fix the conflict. 👍

**Comment by Hoverbear** [comment]
@hicqu Sorry for the delay. :)

File: src/lib.rs [hunk 1]
```diff
@@ -337,8 +336,8 @@ pub mod prelude {
     pub use read_only::{ReadOnlyOption, ReadState};
 }
 
-/// Do any common test initialization. Eg set up logging, setup fail-rs.
-#[cfg(test)]
-fn setup_for_test() {
+/// Do any common test initialization. Eg set up logging.
+#[doc(hidden)]
+pub fn setup_for_test() {
```
Comment (BusyJay) [inline src/lib.rs]: Why not move it to harness?

Comment (Hoverbear) [reply src/lib.rs]: 😲 Good idea.

**Comment by Hoverbear** [comment]
@BusyJay PTAL! :) Resolved.

**Comment by Hoverbear** [comment]
PTAL @hicqu @AndreMouche :)

File: harness/src/interface.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2018 PingCAP, Inc.
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply harness/src/interface.rs]: The PR is from 2018. 🤣

File: harness/src/network.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+use raft::{
+    eraftpb::{Message, MessageType},
+    storage::MemStorage,
+    Config, Raft, Result, NO_LIMIT,
+};
+use rand;
+use std::collections::HashMap;
+
+#[derive(Default, Debug, PartialEq, Eq, Hash)]
+struct Connem {
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply harness/src/network.rs]: I think it is unreasonable to hold up a simple code move that's been waiting for nearly 6 months for documentation. Let's make that a new PR. :)

**Comment by Hoverbear** [comment]
Merging with 2 LGTM from @hicqu and @AndreMouche (they said LGTM but seems it not approve it in the Github UI)

**Pull request merged**